File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/executing/AddEvaluator.java
package org.opencds.cqf.cql.engine.elm.executing;

import java.math.BigDecimal;
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.State;
import org.opencds.cqf.cql.engine.runtime.*;

public class AddEvaluator {
    public static Object add(final Object left, final Object right, final State state) {

        if (left == null || right == null) {
            return null;
        }

        if (left instanceof Integer leftInteger && right instanceof Integer rightInteger) {
            return leftInteger + rightInteger;
        } else if (left instanceof Long leftLong && right instanceof Long rightLong) {
            return leftLong + rightLong;
        } else if (left instanceof BigDecimal leftBigDecimal && right instanceof BigDecimal rightBigDecimal) {
            return Value.verifyPrecision(leftBigDecimal.add(rightBigDecimal), null);
        } else if (left instanceof Quantity leftQuantity && right instanceof Quantity rightQuantity) {
            return UnitConversionHelper.computeWithConvertedUnits(
                    leftQuantity,
                    rightQuantity,
                    (commonUnit, leftValue, rightValue) ->
                            new Quantity().withUnit(commonUnit).withValue(leftValue.add(rightValue)),
                    state);
        }

        // +(DateTime, Quantity), +(Date, Quantity), +(Time, Quantity)
        else if (left instanceof BaseTemporal && right instanceof Quantity rightQuantity) {
            Precision valueToAddPrecision = Precision.fromString(rightQuantity.getUnit());
            Precision precision = Precision.fromString(BaseTemporal.getLowestPrecision((BaseTemporal) left));
            int valueToAdd = rightQuantity.getValue().intValue();

            if (left instanceof DateTime || left instanceof Date) {
                if (valueToAddPrecision == Precision.WEEK) {
                    valueToAdd = TemporalHelper.weeksToDays(valueToAdd);
                    valueToAddPrecision = Precision.DAY;
                }
            }

            if (left instanceof DateTime || left instanceof Date) {
                if (precision == Precision.WEEK) {
                    valueToAdd = TemporalHelper.weeksToDays(valueToAdd);
                    precision = Precision.DAY;
                }
            }
            long convertedValueToAdd = valueToAdd;
            if (precision.toDateTimeIndex() < valueToAddPrecision.toDateTimeIndex()) {
                convertedValueToAdd =
                        TemporalHelper.truncateValueToTargetPrecision(valueToAdd, valueToAddPrecision, precision);
                valueToAddPrecision = precision;
            }

            if (left instanceof DateTime leftDateTime) {
                return new DateTime(
                        leftDateTime.getDateTime().plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit()),
                        precision);
            } else if (left instanceof Date leftDate) {
                return new Date(leftDate.getDate().plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit()))
                        .setPrecision(precision);
            } else {
                return new Time(
                        ((Time) left).getTime().plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit()),
                        precision);
            }
        }

        // +(Uncertainty, Uncertainty)
        else if (left instanceof Interval leftInterval && right instanceof Interval rightInterval) {
            return new Interval(
                    add(leftInterval.getStart(), rightInterval.getStart(), state),
                    true,
                    add(leftInterval.getEnd(), rightInterval.getEnd(), state),
                    true);
        } else if (left instanceof String && right instanceof String) {
            return ((String) left).concat((String) right);
        }

        throw new InvalidOperatorArgument(
                "Add(Integer, Integer), Add(Long, Long), Add(Decimal, Decimal), Add(Quantity, Quantity), Add(Date, Quantity), Add(DateTime, Quantity) or Add(Time, Quantity)",
                String.format(
                        "Add(%s, %s)",
                        left.getClass().getName(), right.getClass().getName()));
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/elm/executing/AddEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import java.math.BigDecimal
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.*

object AddEvaluator {
    @JvmStatic
    fun add(left: Any?, right: Any?): Any? {
        if (left == null || right == null) {
            return null
        }

        if (left is Int && right is Int) {
            return left + right
        }

        if (left is Long && right is Long) {
            return left + right
        } else if (left is BigDecimal && right is BigDecimal) {
            return Value.verifyPrecision(left.add(right), null)
        } else if (left is Quantity && right is Quantity) {
            return Quantity().withValue((left.value)!!.add(right.value)).withUnit(left.unit)
        } else if (left is BaseTemporal && right is Quantity) {
            var valueToAddPrecision = Precision.fromString(right.unit!!)
            var precision = Precision.fromString(BaseTemporal.getLowestPrecision(left))
            var valueToAdd = right.value!!.toInt()

            if (left is DateTime || left is Date) {
                if (valueToAddPrecision == Precision.WEEK) {
                    valueToAdd = TemporalHelper.weeksToDays(valueToAdd)
                    valueToAddPrecision = Precision.DAY
                }
            }

            if (left is DateTime || left is Date) {
                if (precision == Precision.WEEK) {
                    valueToAdd = TemporalHelper.weeksToDays(valueToAdd)
                    precision = Precision.DAY
                }
            }
            var convertedValueToAdd = valueToAdd.toLong()
            if (precision.toDateTimeIndex() < valueToAddPrecision.toDateTimeIndex()) {
                convertedValueToAdd =
                    TemporalHelper.truncateValueToTargetPrecision(
                        valueToAdd.toLong(),
                        valueToAddPrecision,
                        precision,
                    )
                valueToAddPrecision = precision
            }

            if (left is DateTime) {
                return DateTime(
                    left.dateTime!!.plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit()),
                    precision,
                )
            } else if (left is Date) {
                return Date(
                        left.date!!.plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit())
                    )
                    .withPrecision(precision)
            } else {
                return Time(
                    (left as Time)
                        .time
                        .plus(convertedValueToAdd, valueToAddPrecision.toChronoUnit()),
                    precision,
                )
            }
        } else if (left is Interval && right is Interval) {
            val leftInterval = left
            val rightInterval = right
            return Interval(
                add(leftInterval.start, rightInterval.start),
                true,
                add(leftInterval.end, rightInterval.end),
                true,
            )
        } else if (left is String && right is String) {
            return left + right
        }

        throw InvalidOperatorArgument(
            "Add(Integer, Integer), Add(Long, Long), Add(Decimal, Decimal), Add(Quantity, Quantity), Add(Date, Quantity), Add(DateTime, Quantity) or Add(Time, Quantity)",
            String.format("Add(%s, %s)", left.javaClass.name, right.javaClass.name),
        )
    }
}
